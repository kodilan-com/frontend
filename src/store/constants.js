export const TOGGLE_LOADING = 'TOGGLE_LOADING';
export const SET_RECENT_POSTS = 'SET_RECENT_POSTS';
export const SET_ALL_POSTS = 'SET_ALL_POSTS';
export const SET_FEATURED = 'SET_FEATURED';
export const SET_TAGS = 'SET_TAGS';
export const SET_AVAILABLE_LOCATIONS = 'SET_AVAILABLE_LOCATIONS';
export const SET_ACTIVE_PERIOD = 'SET_ACTIVE_PERIOD';
export const RECENT_POST_COUNT = 25;
export const JOB_TYPE_MAP = {
  0: 'Tam zamanlı',
  1: 'Tam zamanlı',
  2: 'Yarı zamanlı',
  3: 'Stajyer',
  4: 'Freelance',
};

export const PERIODS = [
  {
    type: 'daily',
    text: 'Bugün',
    slug: 'bugun',
  },
  {
    type: 'weekly',
    text: 'Bu hafta',
    slug: 'bu-hafta',
  },
  {
    type: 'monthly',
    text: 'Bu ay',
    slug: 'bu-ay',
    isDefault: true,
  },
];

export const JOB_TYPE_SLUG_TO_ID_MAP = {
  'tam-zamanli': 1,
  'yari-zamanli': 2,
  stajyer: 3,
  freelance: 4,
};

export const JOB_TYPE_ID_TO_SLUG_MAP = {
  1: 'tam-zamanli',
  2: 'yari-zamanli',
  3: 'stajyer',
  4: 'freelance',
};

export const JOB_TYPES_FOR_DROPDOWN = [
  {
    id: 1,
    text: 'Tam zamanlı',
  },
  {
    id: 2,
    text: 'Yarı zamanlı',
  },
  {
    id: 3,
    text: 'Stajyer',
  },
  {
    id: 4,
    text: 'Freelance',
  },
];
