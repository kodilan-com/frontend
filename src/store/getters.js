export default {
  autocompleteTags(state) {
    return state.tags.map(tag => ({ label: tag.name, value: tag.slug }));
  },
  companyId(state) {
    return Object.keys(state.me).length && state.me.companies[0].id;
  },
};
