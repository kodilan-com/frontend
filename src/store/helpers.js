export default {
  rankPosts(post, postTagSlugs, relatedPosts) {
    const DAY_IN_MS = 1000 * 60 * 60 * 24;
    const TODAY = new Date();
    const POINTS = {
      createdInTenDays: 2,
      fromSameCompany: 3,
      hasSamePostType: 4,
      hasSameTag: 5,
      inSameLocation: 6,
    };

    return relatedPosts
      .filter(p => p.slug !== post.slug)
      .map((p) => {
        let score = 0;

        score += p.location === post.location ? POINTS.inSameLocation : 0;
        score += p.company.slug === post.company.slug ? POINTS.fromSameCompany : 0;
        score += p.type === post.type ? POINTS.hasSamePostType : 0;
        score += (TODAY - new Date(p.updated_at)) / DAY_IN_MS <= 10 ? POINTS.createdInTenDays : 0;
        p.tags.forEach((t) => {
          score += postTagSlugs.indexOf(t.slug) > -1 ? POINTS.hasSameTag : 0;
        });

        return { score, post: p };
      })
      .sort((c, n) => (c.score > n.score ? -1 : 1))
      .slice(0, 3)
      .map(o => o.post);
  },
};
