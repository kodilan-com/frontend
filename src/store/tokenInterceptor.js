export default (acc) => {
  const token = localStorage.getItem('AccessToken');
  if (token !== null) {
    acc.headers.Authorization = `Bearer ${token}`;
  }
  return acc;
};
