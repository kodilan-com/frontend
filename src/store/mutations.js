import * as constants from './constants';

export default {
  [constants.TOGGLE_LOADING](state) {
    state.isLoading = !state.isLoading;
  },
  [constants.SET_RECENT_POSTS](state, recentPosts) {
    state.recentPosts = recentPosts;
  },
  [constants.SET_ALL_POSTS](state, allPosts) {
    state.allPosts = allPosts;
  },
  [constants.SET_ALL_POST_META](state, meta) {
    state.allPostsMeta = meta;
  },
  [constants.SET_FEATURED](state, featuredPosts) {
    state.featuredPosts = featuredPosts;
  },
  [constants.SET_TAGS](state, tags) {
    state.tags = tags;
  },
  [constants.SET_AVAILABLE_LOCATIONS](state, locations) {
    state.availableLocations = locations.reduce((arr, loc) => {
      if (loc.location !== 'Remote') {
        arr.push(loc.location);
      }

      return arr;
    }, ['Remote']);
  },
  [constants.SET_ACTIVE_PERIOD](state, period) {
    state.activePeriod = period;
  },
  setIsLoggedIn(state, data) {
    state.isLoggedIn = data;
  },
  SET_ME(state, data) {
    state.me = data;
  },
  postListing(state, data) {
    state.manageJobs = data;
  },
};
