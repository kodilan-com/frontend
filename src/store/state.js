import { PERIODS } from './constants';

export default {
  isLoggedIn: localStorage.getItem('AccessToken') || null,
  me: {},
  manageJobs: [],
  isLoading: true,
  recentPosts: [],
  activePeriod: PERIODS.find(period => period.isDefault).type,
  featuredPosts: [],
  allPosts: [],
  allPostsMeta: {},
  tags: [],
  availableLocations: ['Remote', 'İstanbul', 'İzmir', 'Ankara', 'Bursa'],
};
