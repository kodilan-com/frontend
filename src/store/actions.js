import axios from 'axios';
import { cacheAdapterEnhancer } from 'axios-extensions';
import * as constants from './constants';
import helpers from './helpers';

const http = axios.create({
  baseURL: 'https://api.kodilan.com',
  adapter: cacheAdapterEnhancer(axios.defaults.adapter),
});

export default {
  toggleLoading({ commit }) {
    commit(constants.TOGGLE_LOADING);
  },
  fetchRecentPosts({ state, commit }) {
    return http.get(`/posts?get=${constants.RECENT_POST_COUNT}&period=${state.activePeriod}`)
      .then((res) => {
        commit(constants.SET_RECENT_POSTS, res.data.data);

        return res.data;
      });
  },
  fetchFeaturedPosts({ commit }) {
    return http.get('/posts?get=3&is_featured=1')
      .then((res) => {
        commit(constants.SET_FEATURED, res.data.data);

        return res.data;
      });
  },
  fetchAllPosts({ commit }, payload = {}) {
    const page = payload
    && payload.page
    && Number.isInteger(payload.page)
    && payload.page > 1 ? payload.page : 1;

    return http.get(`/posts?get=${constants.PER_PAGE}&page=${page}`)
      .then((res) => {
        const { data } = res;
        commit(constants.SET_ALL_POSTS, data.data);
        commit(constants.SET_ALL_POST_META, {
          total: data.total,
          current_page: data.current_page,
          last_page: data.last_page,
        });

        return res.data;
      });
  },
  fetchBySlug(_, slug) {
    return http.get(`/posts/${slug}`)
      .then(res => res.data);
  },
  fetchByCompany(_, company) {
    return http.get(`/companies/${company}/posts`)
      .then(res => res.data);
  },
  fetchByTag(_, tag) {
    return http.get(`/tags/${tag}/posts`)
      .then(res => res.data);
  },
  search(_, params) {
    return http.get('/search', { params })
      .then(res => res.data);
  },
  fetchTags({ commit }) {
    return http.get('/tags')
      .then((res) => {
        commit(constants.SET_TAGS, res.data.data);
      });
  },
  fetchRelatedPosts({ dispatch }, post) {
    const postTags = post.tags.map(t => t.slug);
    const categoryTags = ['frontend', 'backend', 'mobile', 'designer', 'qa', 'devops'];
    const [mainTag] = postTags.filter(t => categoryTags.indexOf(t) > -1);

    if (!mainTag) {
      return Promise.resolve([]);
    }

    return dispatch('fetchByTag', mainTag)
      .then(res => helpers.rankPosts(post, postTags, res.data));
  },
  fetchAvailableLocations({ commit }) {
    return http.get('/posts/locations')
      .then((res) => {
        commit(constants.SET_AVAILABLE_LOCATIONS, res.data);
      });
  },
  savePost(_, data) {
    return http.post('/posts', data);
  },
  subscribe(_, data) {
    return http.post('/newsletters', data);
  },
  setPeriod({ commit }, period) {
    commit('SET_ACTIVE_PERIOD', period);
  },
};
