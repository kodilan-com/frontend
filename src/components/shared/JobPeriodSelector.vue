<script>
import { mapState, mapActions } from 'vuex';
import { PERIODS } from '../../store/constants';

export default {
  computed: {
    ...mapState(['activePeriod']),
  },
  methods: {
    ...mapActions(['setPeriod']),
  },
  created() {
    this.PERIODS = PERIODS;
  },
};
</script>

<template>
  <ul class="tabs-nav">
    <li
      v-for="period in PERIODS"
      :key="period.type"
      :class="{ active: activePeriod === period.type }"
    >
      <router-link :to="`/${period.slug}`">
        {{ period.text }}
      </router-link>
    </li>
  </ul>
</template>
