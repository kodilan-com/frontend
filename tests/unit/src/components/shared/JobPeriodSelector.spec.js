import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import JobPeriodSelector from '@/components/shared/JobPeriodSelector';
import { PERIODS } from '@/store/constants';

import Vuex from 'vuex';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('JobPeriodSelector.vue', () => {
  let store;
  let wrapper;

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        activePeriod: 'N/A',
      },
      action: {
        setPeriod: jest.fn(),
      },
    });

    wrapper = shallowMount(JobPeriodSelector, {
      stubs: {
        RouterLink: RouterLinkStub,
      },
      store,
      localVue,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('should have 3 periods', () => {
    const periods = wrapper.findAll('ul.tabs-nav li');

    expect(periods.length).toBe(3);
    expect(periods.wrappers.map(period => period.find('a').props('to')))
      .toEqual(PERIODS.map(period => `/${period.slug}`));
    expect(periods.wrappers.map(period => period.find('a').text()))
      .toEqual(PERIODS.map(period => period.text));
  });

  it('should not have active tab', () => {
    expect(wrapper.find('ul.active').exists()).toBe(false);
  });

  it('should have active tab', () => {
    expect(wrapper.find('ul.tabs-nav li.active').exists()).toBe(false);
  });

  it('should have right part', () => {
    const rightPart = wrapper.find('.ul.tabs-nav.is-right');

    expect(rightPart.element).toMatchSnapshot();
    expect(rightPart.find('a').props('to')).toBe('/ilanlar');
  });
});
