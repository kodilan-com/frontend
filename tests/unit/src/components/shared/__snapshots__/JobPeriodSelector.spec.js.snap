// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobPeriodSelector.vue should have right part 1`] = `
<div
  class="ul tabs-nav is-right"
>
  <li
    class="active"
  >
    <a>
      <i
        class="fa fa-plus-circle"
      />
      
        Tüm İlanları Gör
      
    </a>
  </li>
</div>
`;
